Fix filters resetting when returning from session details

R4md4c opened the pull request:
I used rememberSaveable rather than remember in order to preserve the filters across recreations.

Also to be able to save the `SessionFilter`, I noticed that `value` is always a `String`, so I thought it would be appropriate to convert to String in order to make it fully Parcelable. If we wanted to have a more extensible solution for the future, perhaps we could adopt a sealed class approach, where each subtype is a `FilterType`. however, I thought that might be overkill, and let's settle for a simpler approach that does the job decent enough. 

Fixes #121

martinbonnin reviewed (commented):
Could we save the filter state in a view model or higher up in the composition tree? We don't use `rememberSaveable` anywhere else in the app so it might come up as weird if say, the current selected day is lost but not the filter state?

R4md4c commented:
> Could we save the filter state in a view model or higher up in the composition tree? We don't use `rememberSaveable` anywhere else in the app so it might come up as weird if say, the current selected day is lost but not the filter state?

I've taken a look onto your suggestion, but I've met a few issues as a result of how `LceViewModel` is designed.

If added sessionFilters as a MutableStateFlow to the AgendLayoutViewModel, due to how the `produce()` is called in the constructor of the parent class, the `sessionFilters` is still `null` at the time the produce is called hence causing an NPE  

```
class AgendaLayoutViewModel : LceViewModel<Agenda>() {
  private val sessionFilters = MutableStateFlow(emptyList<SessionFilter>()) // This is still null and not accessible from within produce()

  override fun produce(): Flow<Result<AgendaState>> {
    return combine(
        AndroidMakersApplication.instance().store.getAgenda(),
        sessionFilters // NPE Here
    ) {
        // Build out state
    }
  }
}
```

I could see few options on how to proceed: 
1. Move away from LceViewModel to a custom one like in [SessionDetailViewModel](https://github.com/paug/AndroidMakersApp/blob/8ee34fec43f617c712db2dc78b2ec4660b91e9e3/app/src/main/java/fr/paug/androidmakers/ui/components/SessionDetailViewModel.kt#L12-L12)
2. Use `viewModelScope.launch(Dispatchers.Main)` in [here](https://github.com/paug/AndroidMakersApp/blob/8ee34fec43f617c712db2dc78b2ec4660b91e9e3/app/src/main/java/fr/paug/androidmakers/ui/viewmodel/LceViewModel.kt#L23-L23), rather than the regular one that uses on `Dispatchers.Main.immediate`, to give the chance for the child subclasses to initialize itself. 
3. Just host sessionFilters as a separate StateFlow and handle it as a separate `StateFlow` from the [`values`](https://github.com/paug/AndroidMakersApp/blob/8ee34fec43f617c712db2dc78b2ec4660b91e9e3/app/src/main/java/fr/paug/androidmakers/ui/viewmodel/LceViewModel.kt#L14-L14).

WDYT? Open for other suggestions as well.

martinbonnin commented:
Thanks a lot for looking into this! 

> 1. Move away from LceViewModel to a custom one 
> 3.  host sessionFilters as a separate StateFlow and handle it as a separate StateFlow

I think both these suggestions make a lot of sense. `SessionFilter` do not have "LCE" semantics. "LCE" is for network calls that don't have an initial value and can produce an error. In the `SessionFilter` case, we have an initial value (emptyList()) and no error is possible. So I think moving them to their own ViewModel would fix most of the issues.

R4md4c commented:
@martinbonnin just to make sure that I've understood you correctly, do you mean that we move away from LceViewModel and introduce a custom one that just extends `ViewModel` (Option 1)?

martinbonnin commented:
Sorry for the delay, days are a bit hectic these days... I don't want to sound too nitpicky here given that your initial implementation is already much better than what we have today so let me know when this goes into too much bikeshedding 😅 . 
On the other hand, I feel like this is a good opportunity to explore the best practices of where to store state so it's an interesting use case. Right now I can count 3 ways to store state:

1. In the composition tree with `remember {}`
2. In a view model
3. In a bundle with `rememberSaveable`

The scope of the stored data is different each time

1. composition
2. view model
3. state bundle

Back to your original question:

> You mean that we move away from LceViewModel and introduce a custom one that just extends ViewModel

I'd say we can keep LceViewModel for `AgendaLayoutViewModel` and "just" create a new ViewModel to host the filter data? 

`AgendaLayoutViewModel` is clearly a bad name since it is more at the `AgendaPager` level than at the `AgendaLayout` level. So maybe rename `AgendaLayoutViewModel` to `AgendaPagerViewModel` and use `AgendaLayoutViewModel` to host the session filters?

R4md4c commented:
I see, thanks for your elaborate explanation! I think the confusion came from the naming of `AgendaLayoutViewModel` that should've been `AgendaPagerViewModel` as you've mentioned. 

I've implemented your suggestion and now the state is being hosted in a proper `AgendaLayoutViewModel` and I've renamed the other one to `AgendaPagerViewModel`, I've also added a test for the new ViewModel, I've seen no tests in the project so far, so I thought it would be better to start adding one. In case you don't like I can remove it.

martinbonnin commented:
Thanks !

> I've seen no tests in the project so far, so I thought it would be better to start adding one. In case you don't like I can remove it.

Testing is hard 😅 . In general, I wouldn't necessarily test something as simple as the `AgendaLayoutViewModel` but that's a good start 👍 . Let's see where that is taking us :).
